modules/transports/core/vfs/src/main/java/org/apache/synapse/transport/vfs/VFSTransportSender.java

```diff
@@ -119,6 +125,44 @@ public void sendMessage(MessageContext msgCtx, String targetAddress,
             vfsOutInfo = (VFSOutTransportInfo) outTransportInfo;
         }
 
+        WriteLockObject lockObject = null;
+        String baseUri = null;
+        if (vfsOutInfo != null) {
+            if (vfsOutInfo.getSendFileSynchronously()) {
+                baseUri = getBaseUri(targetAddress);
+                if (baseUri != null) {
+                    if (!lockingObjects.containsKey(baseUri)) {
+                        lockingObjects.putIfAbsent(baseUri, new WriteLockObject());
+                        log.debug("New locking object created for Synchronous write|MapSize:" + lockingObjects.size());
+                    }
+                    lockObject = lockingObjects.get(baseUri);
+                    lockObject.incrementUsers();
+                }
+            }
+        }
+
+        try {
+            if (lockObject == null) {
+                writeFile(msgCtx, vfsOutInfo);
+            } else {
+                synchronized (lockObject) {
+                    writeFile(msgCtx, vfsOutInfo);
+                }
+            }
+        }catch (AxisFault axisFault){
+            throw axisFault;
+        }catch (Exception e){
```

Comment:
This was fixed in https://github.com/wso2/wso2-synapse/pull/382 and https://github.com/wso2/wso2-synapse/pull/398